### PR TITLE
Practise 3.1 completed

### PR DESCRIPTION
**Pull request by YaroslavKSE**
Tested new function repeatString. Added 5 more tests for it. Refactored code naming.

### CODE REVIEW

File: Class/Practise 3.1/src/Helpers.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+double dummyFunc(double x) {
+    return x > 0 ? x : 0;
+}
+
+std::string repeatString(const std::vector<int>& vec, const std::string& str) {
+    if (vec.empty()) {
+        return "";
+    }
+    int maxVal = *std::max_element(vec.begin(), vec.end());
```
Comment (vovaskochko) [inline Class/Practise 3.1/src/Helpers.cpp]: std::ranges::max from header <algorithm> should do the trick

and your code can be replaced with something like:
```
    auto repeatCount = vec.empty() ? 0 : std::ranges::max(vec);
    return repeatCount > 0
            ? std::views::repeat(str, repeatCount) | std::views::join | std::ranges::to<std::string>()
            : "";
```

File: Class/Practise 3.1/src/test.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        // Check:
+        ASSERT_EQ(result, 0.0);
+    });
+
+     // Tests for repeatString
+    testSuite.addTest("RepeatStringPrints_home_Twice", []() {
+        std::vector<int> vec = {0, 2, 1};
+        std::string str = "home";
+        auto result = repeatString(vec, str);
```
Comment (vovaskochko) [inline Class/Practise 3.1/src/test.cpp:49]: It is better to add empty lines between logical blocks Build-Operate-Check or Arrange-Act-Assert. Even if you will omit the comment with block name empty lines will make the trick. Otherwise it is harder to read the code.

Also it would be great to have some unit test with empty vector and another one with empty string as an input to cover corner cases

**Review by vovaskochko — COMMENTED** [review]
Looks good except few minor notes:
Formatting of unit tests should be better
It is important to have corner cases covered by unit tests like empty input vector or empty string. Especially if your code has condition check and code branching as it is in the case of empty vector.

File: Class/Practise 3.1/src/Helpers.cpp (hunk 1 above)
Comment (YaroslavKSE) [reply Class/Practise 3.1/src/Helpers.cpp]: Refactored the code in https://github.com/YaroslavKSE/Professinal-Programming/pull/7/commits/eb553f5aca2ab802bbcab07b0f76848c99af5285

File: Class/Practise 3.1/src/test.cpp (hunk 2 above)
Comment (YaroslavKSE) [reply Class/Practise 3.1/src/test.cpp:49]: Refactored code in https://github.com/YaroslavKSE/Professinal-Programming/pull/7/commits/eb553f5aca2ab802bbcab07b0f76848c99af5285

File: Class/Practise 3.1/include/Helpers.hpp [hunk 3]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 #include <vector>
 #include <string>
+#include <ranges>
```
Comment (vovaskochko) [inline Class/Practise 3.1/include/Helpers.hpp:4]: Minor: there is no need to expose this include in header as it is used only inside cpp and you can move it there - as I mentioned on the last lecture - the less includes in translation units the faster compilation is.

File: Class/Practise 3.1/src/Helpers.cpp [hunk 4]
```diff
@@ -7,13 +7,8 @@ double dummyFunc(double x) {
...
-    }
-    int maxVal = *std::max_element(vec.begin(), vec.end());
-    std::string result;
-    for (int i = 0; i < maxVal; ++i) {
-        result += str;
-    }
-    return result;
+    auto repeatCount = vec.empty() ? 0 : std::ranges::max(vec);
+    return repeatCount > 0
+        ? std::views::repeat(str) | std::views::take(repeatCount) | std::views::join | std::ranges::to<std::string>()
```
Comment (vovaskochko) [inline Class/Practise 3.1/src/Helpers.cpp:12]: Infinite lazy views is a great thing - good catch!

**Review by vovaskochko — APPROVED** [review]
